### DISCUSSION TITLE
License as copyleft?

### DISCUSSION THREAD

**Discussion by ThinkChaos** [swingmx/swingmusic#174] [General]
Hi,

Thanks for all the work that's been put into this. I installed it earlier this week to try it out and I'm convinced it could become one of the more popular music servers out there!

There's still some rough edges which I'd love to help soften, but I'd rather not contribute to something MIT licensed because that means my work can then be used by others without them giving back anything to the community.
Would you be open to re-license this as copyleft?  
AGPLv3 seems fitting since this is server software.

I figured I might as well ask since this is a relatively new project, and there are only 2 contributors ATM.  
That could also easily be brought down to just @cwilvx since the second contributor provided the Dockerfile, which I noticed issues with and was thinking could be changed to use alpine instead of ubuntu.

**Comment by michilyy** [comment]
Hello @ThinkChaos ,
I hope you are still active after all this time.

I looked into this topic on [AGPL-3.0](https://choosealicense.com/licenses/agpl-3.0/) and we could change the license.
But I do not have any experience with licenses.
Could you state, which changes this would introduce into the project.

Could you please create a Pull Request in which you change the license.

**Reply by ThinkChaos** [reply]
Hey, thank you for considering this, I'm not very active on any GitHub project ATM, but still here :)

(I'm not a lawyer so all this is not legal advice, but it's my best understanding!)

I'll start with why you would  want AGPL vs current MIT from my POV: MIT basically gives more right to distributors of the software compared to the end users. Where AGPL differs is if you host a modified version of the project, then you must also make the modified source code available to everyone that can use/access your server.
Critically this means if an individual makes modifications for themselves and are the only user, they don't need to publish the changes anywhere since they obviously already have access to them, and they're the only user. So the set of people that have access to the service a superset of the set of people that have access to the code, which is what the AGPL requires. Same thing goes for small groups of people, the source code doesn't necessarily have to be on the open internet if the server is not. As long as it's available to everyone using the service, it's ok.
It also has no effect on people running the service from official releases: that source code is already available to all, so they don't need to republish the source code themselves.
It does require someone/a company from taking the project as a base for a new streaming service without making available any changes they make. So basically if someone creates a Spotify competitor based on Swing, they'll have to publish the source code and the whole Swing community can benefit from the improvements they made.

To transition the project, first I should note that it's not possible to change the past: the MIT license for existing code is not revocable, and only future new code would be impacted. So at any time in the future, someone could go take the MIT version of the source code and fork that without being subject to AGPL.  

The rough steps to transition would be:
1. the license is changed in commit X
2. Existing code that you wrote (which is most of it [according to GitHub](https://github.com/swingmx/swingmusic/graphs/contributors)) is now dual licensed MIT or AGPL
3. Existing code that others wrote stays MIT only
4. New code contributed by anyone is AGPL only.
  As soon as this code is released and used by a service, then it must respect AGPL terms because it's no longer possible to choose only MIT of the dual license (the license is kind of `(MIT or AGPL) and AGPL == AGPL`)
5. Over time more and more code is AGPL only, which protects the project more and more from being forked without giving back, as such a fork would only be able to use the old MIT code, lacking features developed since the license change at commit X.

This is basically [what Forgejo did](https://forgejo.org/2024-08-gpl/).

If you're still interested, I can make a PR!

**Reply by michilyy** [reply]
Hello @ThinkChaos ,
thank you for your comprehensive feedback.
I think I understand you intentions with the license change.

Could you please open a pull request with the changes here and on my fork.
So we can continue developing `swingmusic` until the project owner shows up again.

**Reply by ThinkChaos** [reply]
Sorry for the late reply, I initially thought you were the original maintainer. I'd rather not get involved ATM.

**Reply by michilyy** [reply]
Hello @ThinkChaos ,
The license was already changed.
Your change is no more needed.

Thank you for your involvement.

**Comment by cwilvx** [comment]
Hi @ThinkChaos 

Thank you very much for the license change suggestion. I researched it and updated the license for the project from MIT to AGPLv3.

Thank you very much.
